### PR TITLE
Support parent-based and traceID ratio samplers in eBPF

### PR DESCRIPTION
**Pull request by RonFed**
What is being added:
* `alwaysOn`, `alwaysOff`, `parentBased`, and `traceIDRatio` samplers in eBPF. These are configured in user space by inspecting the `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` environment variables. The configuration is passed to the eBPF code using maps.
* `make test_ebpf` target aims to run all the tests which are (currently) loading maps and (in the future) loading programs. This target needs to be run as a privileged user.

What is removed:
* The `WithSampler` `InstrumentationOption` is removed since it is now misleading and doesn't really act as a sampler since the sampling decision was made after the spans have finished.

### CODE REVIEW

File: instrumentation.go [hunk 1]
```diff
@@ -452,15 +449,6 @@ func WithTraceExporter(exp trace.SpanExporter) InstrumentationOption {
...
 
-// WithSampler returns an [InstrumentationOption] that will configure
-// an [Instrumentation] to use the provided sampler to sample OpenTelemetry traces.
-func WithSampler(sampler trace.Sampler) InstrumentationOption {
-	return fnOpt(func(_ context.Context, c instConfig) (instConfig, error) {
-		c.sampler = sampler
-		return c, nil
-	})
-}
-
```
Comment (MrAlias) [inline instrumentation.go:463]: I think there still needs to be programmatic configuration of the samplers that is going to be used. Likely we need to replace the argument here with some enum like we did for logging.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,9 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 
 ## [Unreleased]
 
+### Added
```
Comment (MrAlias) [inline CHANGELOG.md:11]: An entry for the `WithSampler` option needs to be added.

File: internal/pkg/instrumentation/probe/sampling/sampling.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+// taken from go.opentelemetry.io/otel/sdk/trace and adapted to work with eBPF.
+func (sc *Config) applySamplerFromEnv() error {
+	defaultParentBased := defaultParentBasedSampler()
+	defaultTraceIDRatio := samplerConfig{
+		samplerType: samplerTraceIDRatio,
+		config:      traceIDRationDefaultConfig(),
+	}
+
+	sampler, ok := os.LookupEnv(tracesSamplerKey)
```
Comment (MrAlias) [inline internal/pkg/instrumentation/probe/sampling/sampling.go:209]: This is really low in the program to be interpreting environment variables. Instead, we should have the instrumentation continue to be the place where these are interpreted and have it pass a configuration down to this level.

File: internal/pkg/instrumentation/probe/sampling/sampling.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+// and as a value in the active sampler map. In addition samplers can reference other samplers in their configuration by their ID.
+type samplerID uint32
+
+type parentBasedConfig struct {
+	Root             samplerID
+	RemoteSampled    samplerID
+	RemoteNotSampled samplerID
+	LocalSampled     samplerID
+	LocalNotSampled  samplerID
+	_                [4]byte
```
Comment (MrAlias) [inline internal/pkg/instrumentation/probe/sampling/sampling.go:58]: What is this for?

File: internal/pkg/instrumentation/probe/sampling/sampling.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+// the size of the data is fixed, and the actual configuration is stored in the first part of the data.
+// the rest of the data is padding to make sure the size is fixed.
+
+// The spec-defined samplers have a constant ID, and are always available.
+const (
+	alwaysOnID     samplerID = 0
+	alwaysOffID    samplerID = 1
+	traceIDRatioID samplerID = 2
+	parentBasedID  samplerID = 3
+)
```
Comment (MrAlias) [inline internal/pkg/instrumentation/probe/sampling/sampling.go:95]: These look like duplicates of https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/971/files#diff-066535099a2e1f5733afadd0071ea145869090928742aa0b51daf7a764c54698R33-R40. Why define these twice?

**Comment by RonFed** [comment]
This PR will be split into 2 as discussed in the SIG. Keeping it in draft for now.

**Comment by MrAlias** [comment]
Everything within this PR is merged into main :tada:

**Pull request closed without merging**